### PR TITLE
Return execution exception info properly to session client.

### LINKED ISSUE #505: Return full error messages to client
**Is your feature request related to a problem? Please describe.**
Return full error messages from server side when graph execution failed.

### PR DESCRIPTION
**Pull request by sighingnow**
## What do these changes do?

This pr returns execution exception properly to the client, both for the web session and the localcluster session. After this pr, when meets `ExecutionFailed` we will see (the `Traceback from server` part is what we really want):

```python
>>> x = mt.array(["1", "2", "3", "4"])
>>> sess.run(x + 1)
---------------------------------------------------------------------------
ExecutionFailed                           Traceback (most recent call last)
<ipython-input-7-112e57449079> in <module>
----> 1 sess.run(x + 1)

~/mars/mars/session.py in run(self, *tileables, **kw)
    153         tileables = tuple(mt.tensor(t) if not isinstance(t, (Entity, Base)) else t
    154                           for t in tileables)
--> 155         result = self._sess.run(*tileables, **kw)
    156
    157         for t in tileables:

~/mars/mars/deploy/local/session.py in run(self, *tileables, **kw)
    144                 if exc is not None:
    145                     raise ExecutionFailed('Graph execution failed.\nTraceback from server:\n%s' %
--> 146                         ''.join(traceback.format_exception(*exc)))
    147                 else:
    148                     raise ExecutionFailed('Graph execution failed with unknown reason')

ExecutionFailed: Graph execution failed. Traceback from server:
Traceback (most recent call last):
  File "/Users/hetao/mars/mars/promise.py", line 99, in _wrapped
    result = func(*args, **kwargs)
  File "/Users/hetao/mars/mars/worker/calc.py", line 227, in <lambda>
    .then(lambda context_dict: _start_calc(context_dict)) \
  File "/Users/hetao/mars/mars/worker/calc.py", line 208, in _start_calc
    return self._calc_results(session_id, graph_key, graph, context_dict, chunk_targets)
  File "/Users/hetao/mars/mars/worker/calc.py", line 138, in _calc_results
    chunk_targets, retval=False).result()
  File "src/gevent/event.py", line 344, in gevent._event.AsyncResult.result
  File "src/gevent/event.py", line 268, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 296, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 286, in gevent._event.AsyncResult.get
  File "src/gevent/event.py", line 266, in gevent._event.AsyncResult._raise_exception
  File "/usr/local/lib/python3.7/site-packages/gevent/_compat.py", line 47, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.7/site-packages/gevent/threadpool.py", line 281, in _worker
    value = func(*args, **kwargs)
  File "/Users/hetao/mars/mars/executor.py", line 496, in execute_graph
    res = graph_execution.execute(retval)
  File "/Users/hetao/mars/mars/executor.py", line 389, in execute
    future.result()
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/_base.py", line 435, in result
    return self.__get_result()
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/hetao/mars/mars/executor.py", line 267, in _execute_operand
    Executor.handle(first_op, results, self._mock)
  File "/Users/hetao/mars/mars/executor.py", line 452, in handle
    return runner(results, op)
  File "/Users/hetao/mars/mars/tensor/arithmetic/core.py", line 114, in execute
    ctx[op.outputs[0].key] = _handle_out_dtype(func(lhs, rhs, **kw), op.dtype)
TypeError: ufunc 'add' did not contain a loop with signature matching types dtype('<U3') dtype('<U3') dtype('<U3')
```

## Related issue number

Fixes #505

### CODE REVIEW

File: mars/errors.py [hunk 1]
```diff
@@ -14,13 +14,16 @@
...
+
 
 class MarsError(RuntimeError):
     """
     """
     def __init__(self, msg=None):
         self._err_data = msg
-        msg = repr(msg) if msg is not None else None
+        if msg is not None and not isinstance(msg, six.string_types):
+            msg = repr(msg)
```
Comment (wjsi) [inline mars/errors.py]: Is this used?

File: mars/api.py [hunk 2]
```diff
@@ -163,7 +163,11 @@ def get_graph_state(self, session_id, graph_key):
         except ActorNotExist:
             raise GraphNotExists
         state = GraphState(state.lower())
-        return state
+        if state == GraphState.FAILED:
+            exc = graph_meta_ref.get_exc_info()
+        else:
+            exc = None
+        return state, exc
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply mars/api.py]: Agree, will revise.

File: mars/deploy/local/session.py [hunk 3]
```diff
@@ -133,12 +134,15 @@ def run(self, *tileables, **kw):
...
             self._api.wait_graph_finish(self._session_id, graph_key, timeout=timeout_val)
-            graph_state = self._api.get_graph_state(self._session_id, graph_key)
+            graph_state, exc = self._api.get_graph_state(self._session_id, graph_key)
             if graph_state == GraphState.SUCCEEDED:
                 break
             if graph_state == GraphState.FAILED:
-                # TODO(qin): add traceback
-                raise ExecutionFailed('Graph execution failed with unknown reason')
+                if exc is not None:
+                    raise ExecutionFailed('Graph execution failed. Traceback from server:\n%s' %
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply mars/deploy/local/session.py]: This suggestion looks really promising, will fix soon.

File: mars/errors.py (hunk 1 above)
Comment (sighingnow) [reply mars/errors.py]: This modification is aimed to yield better error message. The `repr('abc')` will become`"'abc'"`. For example:

```python
In [2]: raise Exception('abcd\n\tdef')
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-2-fc45df11d92c> in <module>
----> 1 raise Exception('abcd\n\tdef')

Exception: abcd
	def

In [3]: raise Exception(repr('abcd\n\tdef'))
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-3-51832293106b> in <module>
----> 1 raise Exception(repr('abcd\n\tdef'))

Exception: 'abcd\n\tdef'
```

But if we revise the `ExecutionFailed` part according to the comment above, I think this optimization can be removed.

File: mars/worker/storage/sharedstore.py [hunk 4]
```diff
@@ -14,6 +14,13 @@
 
 import logging
 
+import pyarrow
+from pyarrow import plasma
+try:
+    from pyarrow.plasma import PlasmaObjectNonexistent, PlasmaStoreFull
+except ImportError:
+    from pyarrow.lib import PlasmaObjectNonexistent, PlasmaStoreFull
+
```
Comment (wjsi) [inline mars/worker/storage/sharedstore.py]: Errors caused by pyarrow 0.15.0 has already been fixed in #771.

Comment (sighingnow) [reply mars/worker/storage/sharedstore.py]: Rebased.

File: mars/errors.py (hunk 1 above)
Comment (wjsi) [reply mars/errors.py]: Maybe it is better to put this logic into __str__ to reduce the cost of ``isinstance``.

File: mars/api.py [hunk 5]
```diff
@@ -162,8 +162,14 @@ def get_graph_state(self, session_id, graph_key):
             state = state_obj.value if state_obj else 'preparing'
         except ActorNotExist:
             raise GraphNotExists
-        state = GraphState(state.lower())
-        return state
+        return GraphState(state.lower())
+
+    def get_graph_traceback(self, session_id, graph_key):
```
Comment (wjsi) [inline mars/api.py]: Maybe ``get_graph_exc_info`` is better, as exc_info contains exc_type, exc and traceback.

File: mars/web/apihandlers.py [hunk 6]
```diff
@@ -147,18 +148,12 @@ def _wait_fun():
...
-            self.write(json.dumps(dict(state='cancelled')))
-        elif state == GraphState.CANCELLING:
-            self.write(json.dumps(dict(state='cancelling')))
-        elif state == GraphState.PREPARING:
-            self.write(json.dumps(dict(state='preparing')))
+        if state == GraphState.FAILED:
+            exc = self.web_api.get_graph_traceback(session_id, graph_key)
+            if exc is not None:
+                self.write(json.dumps(dict(state=state.value,
+                                           traceback=''.join(traceback.format_exception(*exc)))))
```
Comment (wjsi) [inline mars/web/apihandlers.py]: I wonder if you can put pickled traceback data here if possible to unify the behavior of actor clients and web clients? String traceback can be a backup.

**Comment by sighingnow** [comment]
How could I fix the security warning about the usage of `pickle` on codacy?

**Comment by qinxuye** [comment]
> How could I fix the security warning about the usage of `pickle` on codacy?

Ignored already:)

File: mars/web/apihandlers.py (hunk 6 above)
Comment (sighingnow) [reply mars/web/apihandlers.py]: Fixed. Now the web session client return both the pickled `exc_info` and the textual traceback.

File: mars/api.py (hunk 5 above)
Comment (sighingnow) [reply mars/api.py]: Fixed.

File: mars/deploy/local/tests/test_cluster.py [hunk 7]
```diff
@@ -120,6 +120,20 @@ def testLocalClusterWithWeb(self, *_):
                              [' '.join(p.cmdline()) for p in processes if p.is_running()])
                 self.assertFalse(any(p.is_running() for p in processes))
 
+    def testLocalClusterError(self, *_):
+        with option_context({'scheduler.retry_num': 1}):
+            with new_cluster(scheduler_n_process=2, worker_n_process=3,
+                            shared_memory='20M', web=True) as cluster:
+                with cluster.session as session:
+                    t = mt.array(["1", "2", "3", "4"])
+                    with self.assertRaises(ExecutionFailed):
```
Comment (qinxuye) [inline mars/deploy/local/tests/test_cluster.py]: Only error type tested, and I guess even error_info is not added,  the ut can still pass? How about adding some tests about the traceback?

File: mars/deploy/local/tests/test_cluster.py [hunk 8]
```diff
@@ -120,6 +120,20 @@ def testLocalClusterWithWeb(self, *_):
...
+                        session.run(t + 1)
+
+                with new_session('http://' + cluster._web_endpoint) as session:
+                    t = mt.array(["1", "2", "3", "4"])
+                    with self.assertRaises(ExecutionFailed):
```
Comment (qinxuye) [inline mars/deploy/local/tests/test_cluster.py]: Ditto.

**Comment by qinxuye** [comment]
BTW, The web api could be tested as well.

File: mars/deploy/local/tests/test_cluster.py (hunk 7 above)
Comment (sighingnow) [reply mars/deploy/local/tests/test_cluster.py]: The validation for the nested exception message has been added.

File: mars/tests/test_api.py [hunk 9]
```diff
@@ -62,10 +62,14 @@ def testApi(self, *_):
 
         state = self.api.get_graph_state(session_id, graph_key)
         self.assertEqual(GraphState('preparing'), state)
+        exc_info = self.api.get_graph_exc_info(session_id, graph_key)
+        self.assertEqual(None, exc_info)
```
Comment (qinxuye) [inline mars/tests/test_api.py]: Use `self.assertIsNone` instead.

File: mars/tests/test_api.py [hunk 10]
```diff
@@ -62,10 +62,14 @@ def testApi(self, *_):
...
 
         self.api.stop_graph(session_id, graph_key)
         state = self.api.get_graph_state(session_id, graph_key)
         self.assertEqual(GraphState('cancelled'), state)
+        exc_info = self.api.get_graph_exc_info(session_id, graph_key)
+        self.assertEqual(None, exc_info)
```
Comment (qinxuye) [inline mars/tests/test_api.py]: Ditto.

**Pull request merged**